File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/backup/stages/internal/PartitionedSource.kt
package stasis.client_android.lib.ops.backup.stages.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.Buffer
import okio.BufferedSink
import okio.BufferedSource
import okio.Sink
import okio.buffer
import okio.sink
import stasis.client_android.lib.encryption.secrets.DeviceFileSecret
import stasis.client_android.lib.ops.backup.Providers
import java.nio.file.Path
import kotlin.math.min

class PartitionedSource(
    private val source: BufferedSource,
    private val providers: Providers,
    private val withPartSecret: (Int) -> DeviceFileSecret,
    private val onPartStaged: () -> Unit,
    private val withMaximumPartSize: Long
) {
    suspend fun partitionAndStage(): List<Pair<Path, Path>> = withContext(Dispatchers.IO) {
        val parts = mutableListOf<Pair<Path, Path>>()

        var collected: Long = 0L
        var partId: Int = 0
        var secret: DeviceFileSecret = withPartSecret(partId)
        var staged: BufferedSink = providers.staging.temporary().let { temporary ->
            parts += secret.file to temporary
            temporary.sink().encrypted(providers, secret).buffer()
        }

        val next = Buffer()

        try {
            while (!source.exhausted()) {
                val read = source.read(next, min(withMaximumPartSize, SourceMaxChunkSize))

                if (read > 0) {
                    if (collected + read > withMaximumPartSize) {
                        staged.flush()
                        staged.flush()
                        staged.close()
                        onPartStaged()

                        collected = read
                        partId += 1
                        secret = withPartSecret(partId)
                        staged = providers.staging.temporary().let { temporary ->
                            parts += secret.file to temporary
                            temporary.sink().encrypted(providers, secret).buffer()
                        }
                    } else {
                        collected += read
                    }

                    staged.write(next, read)
                }
            }
        } catch (e: Throwable) {
            parts.forEach { (_, staged) -> providers.staging.discard(staged) }
            throw e
        } finally {
            staged.flush()
            staged.close()
            onPartStaged()

            source.close()
        }

        parts
    }

    private fun Sink.encrypted(providers: Providers, fileSecret: DeviceFileSecret): Sink =
        providers.encryptor.encrypt(this, fileSecret)

    companion object {
        const val SourceMaxChunkSize: Long = 8192L
    }
}

File: client-ios/lib/Sources/StasisClientLib/Ops/Backup/Internal/PartitionedSource.swift
import Foundation

public struct PartitionedSource: Sendable {
    public typealias Source = AsyncThrowingStream<Data, Error>

    private let source: Source
    private let providers: BackupProviders
    private let withPartSecret: @Sendable (Int) -> DeviceFileSecret
    private let onPartStaged: @Sendable () -> Void
    private let maximumPartSize: Int64

    public init(
        source: Source,
        providers: BackupProviders,
        withPartSecret: @escaping @Sendable (Int) -> DeviceFileSecret,
        onPartStaged: @escaping @Sendable () -> Void,
        maximumPartSize: Int64
    ) {
        self.source = source
        self.providers = providers
        self.withPartSecret = withPartSecret
        self.onPartStaged = onPartStaged
        self.maximumPartSize = maximumPartSize
    }

    public func partitionAndStage() async throws -> [(file: String, path: URL)] {
        var parts: [(file: String, path: URL)] = []
        var collected = Data()
        var partId = 0
        var secret = withPartSecret(partId)
        var temporary = try await providers.staging.temporary()
        parts.append((secret.file, temporary))

        do {
            for try await chunk in source {
                var idx = 0
                while idx < chunk.count {
                    let space = Int(maximumPartSize) - collected.count
                    if space == 0 {
                        try await flushPart(&collected, secret: secret, to: temporary)
                        onPartStaged()

                        partId += 1
                        secret = withPartSecret(partId)
                        temporary = try await providers.staging.temporary()
                        parts.append((secret.file, temporary))
                        continue
                    }

                    let take = min(space, chunk.count - idx)
                    let start = chunk.startIndex.advanced(by: idx)
                    let end = chunk.startIndex.advanced(by: idx + take)
                    collected.append(chunk[start..<end])
                    idx += take
                }
            }

            try await flushPart(&collected, secret: secret, to: temporary)
            onPartStaged()
            return parts
        } catch {
            for (_, path) in parts {
                try? await providers.staging.discard(file: path)
            }
            onPartStaged()
            throw error
        }
    }

    private func flushPart(_ collected: inout Data, secret: DeviceFileSecret, to temporary: URL) async throws {
        let ciphertext = try providers.encryptor.encrypt(collected, fileSecret: secret)
        try ciphertext.write(to: temporary)
        collected = Data()
    }
}
